ext/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -99,222 +103,414 @@ accelerators while being able to synchronize and optimize data transfers at
-. int: Rounding method for calculating output dimensions. 
-. int: A VX_TYPE_ENUM of the vx_convert_policy_e enumeration.
-. size_t: Number of elements padded at each side in the x dimension of the input.
-. size_t: Number of elements padded at each side in the y dimension of the input.
-. int:  A VX_TYPE_ENUM of the vx_round_policy_e enumeration.
-. uchar* [out]: The output tensor data. Output will have the same number and structure of dimensions as input. Output tensor data type must be same as the inputs. (Kernel parameter #4)
-
-4+| OpenCL C Semantics
-4+a|
-[source,c]
-----
-__kernel void __khr_openvx_nn_extension_convolution_uchar(
-   const uchar *input, const uchar *weights, const uchar *biases,
-   size_t dilation_x, size_t dilation_y,
-   int down_scale_rounding, int overflow_policy, size_t padding_x, size_t padding_y,
-   int rounding_policy, uchar *output) {
-   // TBD.
-}
-----
+Performs (batched) matrix multiplication: `R = trans(A) * trans(B)`,
+where `A`, `B` and `R` are tensors with at least rank two. The
+`trans()` is a configurable transpose operation.
+
+Last two dimensions of the tensors are treated as operands to the
+matric multiplication and rest of the dimensions are treated as batch
+dimensions.
+
+Operations of the matrix muliplication are performed in the precision
+of the `elementof\(R)`.
 
-4+| *Direct Input/Output Operations*
-4+| Kernels for accessing data sources and destinations directly without host involvement.
-| Name | Description | NDRange Dimensions | Arguments
-| *khr_io_stream_in_uchar*
-| Non-blocking read of data from a sensor/stream associated with the device.
-a| -
+If an overflow occurs in the accumulation of the products, then `R`
+tensor's result will be undefined.
+
+| *Attribute validation rules*
 a|
-[start=0]
-. uchar* [out]: The data.
-. size_t* [in+out]: In: number of bytes to read. Out: Number of bytes that could be read (can be 0). (Compatible with the `cl_pocl_content_size` extension to optimize data transfers with.)
 
-4+| OpenCL C Semantics
-4+a|
-[source,c]
-----
-__kernel void __khr_io_stream_in_uchar(
-   uchar *output, size_t *num) {
-   // It is not feasible to describe this kernel in OpenCL C as I/O devices
-   // are not representable with it.
-}
-----
+* `rankof(A) == rankof(B) >= 2`.
+* Let `shapeof(A~t~) == (b..., m, k)` and `shapeof(B~t~) = (b..., k,
+  n)` of tensors `A` and `B`, respectively, after possible tranposing.
+  `shapeof\(R)` must be `(b..., m, n)`.
+* `elementof(A) == elementof(B)`
+* `elemkindof\(R) == elemkindof(A)`
+* `elementof\(R) == elementof(A)` or `elementof(A)` is promotable to
+  `elementof\(R)` without loss of meaning.
+// E.g. cl_int -> cl_uint: loses negative values
+| *Kernel mode properties*
+a|
+This DBK accepts the following properties:
 
-| *khr_io_stream_out_uchar*
-| Non-blocking write of data to an output/sink associated with the device.
-| -
+* CL_DBK_PROPERTY_MAX_RELATIVE_ERROR: Unset property defaults to positive infinity.
```

Comment:
Perhaps we should describe the properties elsewhere as most of them should be generic and usable across multiple DBKs. Then here we could just list them by name.

Reply:
Properties are already described elsewhere and already are generic and reusable (maybe not clearly). This DBK just names the properties with default values.